src/sage/modular/drinfeld_modform/ring.py

```diff
@@ -0,0 +1,767 @@
+            ...
+            ValueError: base ring characteristic must be finite
+
+            sage: R = GF(2)['u']
+            sage: K = Frac(R['T'])
+            sage: DrinfeldModularForms(K, 2)
+            Traceback (most recent call last):
+            ...
+            ValueError: the ring of constants must be a field
+
+            sage: K = Frac(Frac(R)['T'])
+            sage: DrinfeldModularForms(K, 2)
+            Traceback (most recent call last):
+            ...
+            ValueError: the ring of constants must be finite
+
+            sage: DrinfeldModularForms(Frac(GF(2)['T']), group=2)
+            Traceback (most recent call last):
+            ...
+            NotImplementedError: Drinfeld modular forms are currently only implemented for the full group
+
+            sage: DrinfeldModularForms(Frac(GF(2)['T']), names=1)
+            Traceback (most recent call last):
+            ...
+            TypeError: names must be None, a comma seperated string or a list of string
+
+            sage: DrinfeldModularForms(Frac(GF(2)['T']), rank=3, names='f1, f2, f3, f4')
+            Traceback (most recent call last):
+            ...
+            ValueError: the number of generators (=4) must be equal to the rank (=3)
+
+            sage: DrinfeldModularForms(Frac(GF(2)['T']), rank=3, names=['f1', 'f2', 'f3', 'f4'])
+            Traceback (most recent call last):
+            ...
+            ValueError: the number of generators (=4) must be equal to the rank (=3)
+
+            sage: DrinfeldModularForms(Frac(GF(2)['T']), rank=3, names=('f1', 'f2', 'f3', 'f4'))
+            Traceback (most recent call last):
+            ...
+            ValueError: the number of generators (=4) must be equal to the rank (=3)
+
+            sage: DrinfeldModularForms(Frac(GF(2)['T']))
+            Traceback (most recent call last):
+            ...
+            TypeError: rank or names must be specified
+        """
+        if not isinstance(base_ring, FractionField_generic):
+            raise TypeError("base ring must be a fraction field of a "
+                            "polynomial ring")
+        if not isinstance(base_ring.base(), PolynomialRing_general):
+            raise NotImplementedError("Drinfeld modular forms are currently "
+                                      "only implemented for A = Fq[T]")
+        if not base_ring.characteristic():
+            raise ValueError("base ring characteristic must be finite")
+        if not base_ring.base().base().is_field():
+            raise ValueError("the ring of constants must be a field")
+        if not base_ring.base().base().is_finite():
+            raise ValueError("the ring of constants must be finite")
+        if group is not None:  # placeholder
+            raise NotImplementedError("Drinfeld modular forms are currently "
+                                      "only implemented for the full group")
+        if names is None: # default names
+            if rank is None:
```

Comment:
There is no doctest corresponding to this branch, I think.

Reply:
Yes there is one, see the [line 259](https://github.com/DavidAyotte/sage/blob/drinfeld_modform/src/sage/modular/drinfeld_modform/ring.py#L259).

Reply:
Oh sorry, I missed it.